pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/StrictRealtimeSegmentAssignment.java

```diff
@@ -0,0 +1,108 @@
+ *     InstancePartition and the one in idealState are different, the one in idealState must be used so that segments
+ *     from the same table partition are always hosted on the same server as set in current idealState. If the
+ *     idealState is not honored, segments from the same table partition may be assigned to different servers,
+ *     breaking the key assumption for queries to be correct for the table using upsert.
+ *   </li>
+ *   <li>
+ *     There is no need to handle COMPLETED segments for tables using upsert, because their completed segments should
+ *     not be relocated to servers tagged to host COMPLETED segments. Basically, upsert-enabled tables can only use
+ *     servers tagged for CONSUMING segments to host both consuming and completed segments from a table partition.
+ *   </li>
+ * </ul>
+ */
+public class StrictRealtimeSegmentAssignment extends RealtimeSegmentAssignment {
+
+  @Override
+  public List<String> assignSegment(String segmentName, Map<String, Map<String, String>> currentAssignment,
+      Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap) {
+    Preconditions.checkState(instancePartitionsMap.size() == 1, "One instance partition type should be provided");
+    Map.Entry<InstancePartitionsType, InstancePartitions> typeToInstancePartitions =
+        instancePartitionsMap.entrySet().iterator().next();
+    InstancePartitionsType instancePartitionsType = typeToInstancePartitions.getKey();
+    InstancePartitions instancePartitions = typeToInstancePartitions.getValue();
+    Preconditions.checkState(instancePartitionsType == InstancePartitionsType.CONSUMING,
+        "Only CONSUMING instance partition type is allowed for table using upsert but got: " + instancePartitionsType);
+    _logger.info("Assigning segment: {} with instance partitions: {} for table: {}", segmentName, instancePartitions,
+        _tableNameWithType);
+    int segmentPartitionId =
+        SegmentAssignmentUtils.getRealtimeSegmentPartitionId(segmentName, _tableNameWithType, _helixManager,
+            _partitionColumn);
+    List<String> instancesAssigned = assignConsumingSegment(segmentPartitionId, instancePartitions);
+    List<String> idealAssignment = new ArrayList<>(instancesAssigned.size());
+    // Iterate the idealState to find the first segment that's in the same table partition with the new segment, and
+    // check if their assignments are same. We try to derive the partition id from segment name to avoid ZK reads.
+    Set<String> nonStandardSegments = new HashSet<>();
+    for (Map.Entry<String, Map<String, String>> entry : currentAssignment.entrySet()) {
+      LLCSegmentName llcSegmentName = LLCSegmentName.of(entry.getKey());
+      if (llcSegmentName == null) {
+        nonStandardSegments.add(entry.getKey());
+        continue;
+      }
+      if (llcSegmentName.getPartitionGroupId() == segmentPartitionId) {
+        idealAssignment.addAll(entry.getValue().keySet());
```

Comment:
We need to skip the segments with all state as `OFFLINE` because they won't be rebalanced

Reply:
Make sense